File: python/problems/pg42584/solution.py
"""
[PG] 42584 - 주식가격
https://programmers.co.kr/learn/courses/30/lessons/42584
난이도: lv2
태그: stack,queue

시간복잡도: O(N): for문 N, stack -> push, pop N, 2N
공간복잡도: O(N)
"""


def solution(prices: list[int]) -> list[int]:
    stack = []
    n = len(prices)
    answer = [0]*n
    for i, price in enumerate(prices):
        while stack and prices[stack[-1]] > price:
            stack_pop = stack.pop()
            answer[stack_pop] = i - stack_pop
        stack.append(i)
    while stack:
        stack_pop = stack.pop()
        answer[stack_pop] = n-1 - stack_pop
    return answer

File: java/src/main/java/problems/pg42584/Solution.java
package problems.pg42584;

import java.util.ArrayDeque;
import java.util.Deque;

/**
 * [PG] 42584 - 주식가격
 * https://programmers.co.kr/learn/courses/30/lessons/42584
 * 난이도: lv2
 * 태그: stack,queue
 *
 * 시간복잡도: O(n)
 * 공간복잡도: O(n)
 */
public class Solution {

    public int[] solution(int[] prices) {
        int n = prices.length;
        // 초기 용량을 설정하여 내부 배열 재할당 방지
        Deque<Integer> stack = new ArrayDeque<>(n);
        int[] answer = new int[n];

        for (int i = 0; i < prices.length; i++) {
            while (!stack.isEmpty() && prices[stack.peekLast()] > prices[i]) {
                final int pollLast = stack.pollLast();
                answer[pollLast] = i - pollLast;
            }
            stack.addLast(i);
        }
        while (!stack.isEmpty()) {
            final int pollLast = stack.pollLast();
            answer[pollLast] = n - 1 - pollLast;
        }
        return answer;
    }
}
